many runaway errors in log

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

checked any log

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

for every second

```
[2024/03/23 10:25:46.302 +08:00] [ERROR] [runaway.go:147] ["try to get new runaway watch"] [error="[schema:1146]Table 'mysql.tidb_runaway_watch' doesn't exist"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/domain/domain.go
+++ b/pkg/domain/domain.go
@@ -1290,8 +1290,7 @@ func (do *Domain) Init(
 	do.wg.Run(do.topNSlowQueryLoop, "topNSlowQueryLoop")
 	do.wg.Run(do.infoSyncerKeeper, "infoSyncerKeeper")
 	do.wg.Run(do.globalConfigSyncerKeeper, "globalConfigSyncerKeeper")
-	do.wg.Run(do.runawayRecordFlushLoop, "runawayRecordFlushLoop")
-	do.wg.Run(do.runawayWatchSyncLoop, "runawayWatchSyncLoop")
+	do.wg.Run(do.runawayStartLoop, "runawayStartLoop")
 	do.wg.Run(do.requestUnitsWriterLoop, "requestUnitsWriterLoop")
 	if !skipRegisterToDashboard {
 		do.wg.Run(do.topologySyncerKeeper, "topologySyncerKeeper")

--- a/pkg/domain/resourcegroup/runaway.go
+++ b/pkg/domain/resourcegroup/runaway.go
@@ -185,6 +185,7 @@ func (r *QuarantineRecord) GenDeletionStmt() (string, []any) {
 // RunawayManager is used to detect and record runaway queries.
 type RunawayManager struct {
 	syncerInitialized atomic.Bool
+	logOnce           sync.Once
 
 	// queryLock is used to avoid repeated additions. Since we will add new items to the system table,
 	// in order to avoid repeated additions, we need a lock to ensure that
@@ -291,10 +292,9 @@ func (rm *RunawayManager) markQuarantine(resourceGroupName, convict string, watc
 	// Add record without ID into watch list in this TiDB right now.
 	rm.addWatchList(record, ttl, false)
 	if !rm.syncerInitialized.Load() {
-		logutil.BgLogger().Warn("runaway syncer is not initialized, so can't record runaway watch",
-			zap.String("resource group", resourceGroupName),
-			zap.String("watch-text", convict),
-			zap.String("watch type", watchType.String()))
+		rm.logOnce.Do(func() {
+			logutil.BgLogger().Warn("runaway syncer is not initialized, so can't records about runaway")
+		})
 		return
 	}
 	select {
@@ -396,9 +396,9 @@ func (rm *RunawayManager) getWatchFromWatchList(key string) *QuarantineRecord {
 func (rm *RunawayManager) markRunaway(resourceGroupName, originalSQL, planDigest string, action string, matchType RunawayMatchType, now *time.Time) {
 	source := rm.serverID
 	if !rm.syncerInitialized.Load() {
-		logutil.BgLogger().Warn("runaway syncer is not initialized, so can't record runaway queries",
-			zap.String("resource group", resourceGroupName),
-			zap.String("sql", originalSQL))
+		rm.logOnce.Do(func() {
+			logutil.BgLogger().Warn("runaway syncer is not initialized, so can't records about runaway")
+		})
 		return
 	}
 	select {

--- a/pkg/domain/resourcegroup/runaway.go
+++ b/pkg/domain/resourcegroup/runaway.go
@@ -184,6 +184,8 @@ func (r *QuarantineRecord) GenDeletionStmt() (string, []any) {
 
 // RunawayManager is used to detect and record runaway queries.
 type RunawayManager struct {
+	syncerInitialized atomic.Bool
+
 	// queryLock is used to avoid repeated additions. Since we will add new items to the system table,
 	// in order to avoid repeated additions, we need a lock to ensure that
 	// action "judging whether there is this record in the current watch list and adding records" have atomicity.
@@ -219,6 +221,7 @@ func NewRunawayManager(resourceGroupCtl *rmclient.ResourceGroupsController, serv
 	go watchList.Start()
 	staleQuarantineChan := make(chan *QuarantineRecord, maxWatchRecordChannelSize)
 	m := &RunawayManager{
+		syncerInitialized:     atomic.Bool{},
 		resourceGroupCtl:      resourceGroupCtl,
 		watchList:             watchList,
 		serverID:              serverAddr,
@@ -243,6 +246,10 @@ func NewRunawayManager(resourceGroupCtl *rmclient.ResourceGroupsController, serv
 		staleQuarantineChan <- i.Value()
 	})
 	return m
+}
+
+func (rm *RunawayManager) MarkSyncerInitialized() {
+	rm.syncerInitialized.Store(true)
 }
 
 // DeriveChecker derives a RunawayChecker from the given resource group
@@ -282,6 +289,13 @@ func (rm *RunawayManager) markQuarantine(resourceGroupName, convict string, watc
 	}
 	// Add record without ID into watch list in this TiDB right now.
 	rm.addWatchList(record, ttl, false)
+	if !rm.syncerInitialized.Load() {
+		logutil.BgLogger().Warn("runaway syncer is not initialized, so can't record runaway watch",
+			zap.String("resource group", resourceGroupName),
+			zap.String("watch-text", convict),
+			zap.String("watch type", watchType.String()))
+		return
+	}
 	select {
 	case rm.quarantineChan <- record:
 	default:
@@ -380,6 +394,12 @@ func (rm *RunawayManager) getWatchFromWatchList(key string) *QuarantineRecord {
 
 func (rm *RunawayManager) markRunaway(resourceGroupName, originalSQL, planDigest string, action string, matchType RunawayMatchType, now *time.Time) {
 	source := rm.serverID
+	if !rm.syncerInitialized.Load() {
+		logutil.BgLogger().Warn("runaway syncer is not initialized, so can't record runaway queries",
+			zap.String("resource group", resourceGroupName),
+			zap.String("sql", originalSQL))
+		return
+	}
 	select {
 	case rm.runawayQueriesChan <- &RunawayRecord{
 		ResourceGroupName: resourceGroupName,

--- a/pkg/domain/resourcegroup/runaway.go
+++ b/pkg/domain/resourcegroup/runaway.go
@@ -248,6 +248,7 @@ func NewRunawayManager(resourceGroupCtl *rmclient.ResourceGroupsController, serv
 	return m
 }
 
+// MarkSyncerInitialized is used to mark the syncer is initialized.
 func (rm *RunawayManager) MarkSyncerInitialized() {
 	rm.syncerInitialized.Store(true)
 }

--- a/pkg/domain/runaway.go
+++ b/pkg/domain/runaway.go
@@ -162,12 +162,12 @@ func (do *Domain) runawayStartLoop() {
 				return
 			}
 		}
-		count++
 		if count %= runawayLoopLogErrorIntervalCount; count == 0 {
 			logutil.BgLogger().Warn(
 				"failed to start runaway manager loop, please check whether the bootstrap or update is finished",
 				zap.Error(err))
 		}
+		count++
 	}
 }
 
@@ -202,10 +202,10 @@ func (do *Domain) runawayWatchSyncLoop() {
 		case <-runawayWatchSyncTicker.C:
 			err := do.updateNewAndDoneWatch()
 			if err != nil {
-				count++
 				if count %= runawayLoopLogErrorIntervalCount; count == 0 {
 					logutil.BgLogger().Warn("get runaway watch record failed", zap.Error(err))
 				}
+				count++
 			}
 		}
 	}

--- a/pkg/domain/runaway.go
+++ b/pkg/domain/runaway.go
@@ -54,8 +54,8 @@ const (
 	runawayRecordGCBatchSize       = 100
 	runawayRecordGCSelectBatchSize = runawayRecordGCBatchSize * 5
 
-	maxIDRetries                = 3
-	runawayLoopLogErrorInterval = 5 * time.Minute
+	maxIDRetries                     = 3
+	runawayLoopLogErrorIntervalCount = 1800
 )
 
 var systemSchemaCIStr = model.NewCIStr("mysql")
@@ -146,9 +146,7 @@ func (do *Domain) runawayStartLoop() {
 	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
 	count := 0
 	var err error
-	logutil.BgLogger().Info(
-		"try to start runaway manager loop",
-		zap.Error(err))
+	logutil.BgLogger().Info("try to start runaway manager loop")
 	for {
 		select {
 		case <-do.exit:
@@ -157,6 +155,7 @@ func (do *Domain) runawayStartLoop() {
 			// Due to the watch and watch done tables is created later than runaway queries table
 			err = do.updateNewAndDoneWatch()
 			if err == nil {
+				logutil.BgLogger().Info("preparations for the runaway manager are finished and start runaway manager loop")
 				do.wg.Run(do.runawayRecordFlushLoop, "runawayRecordFlushLoop")
 				do.wg.Run(do.runawayWatchSyncLoop, "runawayWatchSyncLoop")
 				do.runawayManager.MarkSyncerInitialized()
@@ -164,7 +163,7 @@ func (do *Domain) runawayStartLoop() {
 			}
 		}
 		count++
-		if count == 1800 {
+		if count %= runawayLoopLogErrorIntervalCount; count == 0 {
 			logutil.BgLogger().Warn(
 				"failed to start runaway manager loop, please check whether the bootstrap or update is finished",
 				zap.Error(err))
@@ -195,7 +194,7 @@ func (do *Domain) updateNewAndDoneWatch() error {
 func (do *Domain) runawayWatchSyncLoop() {
 	defer util.Recover(metrics.LabelDomain, "runawayWatchSyncLoop", nil, false)
 	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
-	lastLogErrorTime := time.Now()
+	count := 0
 	for {
 		select {
 		case <-do.exit:
@@ -203,9 +202,9 @@ func (do *Domain) runawayWatchSyncLoop() {
 		case <-runawayWatchSyncTicker.C:
 			err := do.updateNewAndDoneWatch()
 			if err != nil {
-				if now := time.Now(); now.Sub(lastLogErrorTime) > runawayLoopLogErrorInterval {
+				count++
+				if count %= runawayLoopLogErrorIntervalCount; count == 0 {
 					logutil.BgLogger().Warn("get runaway watch record failed", zap.Error(err))
-					lastLogErrorTime = now
 				}
 			}
 		}

--- a/pkg/domain/runaway.go
+++ b/pkg/domain/runaway.go
@@ -54,7 +54,8 @@ const (
 	runawayRecordGCBatchSize       = 100
 	runawayRecordGCSelectBatchSize = runawayRecordGCBatchSize * 5
 
-	maxIDRetries = 3
+	maxIDRetries                = 3
+	runawayLoopLogErrorInterval = 5 * time.Minute
 )
 
 var systemSchemaCIStr = model.NewCIStr("mysql")
@@ -140,12 +141,42 @@ func (do *Domain) deleteExpiredRows(tableName, colName string, expiredDuration t
 	}
 }
 
+func (do *Domain) runawayStartLoop() {
+	defer util.Recover(metrics.LabelDomain, "runawayStartLoop", nil, false)
+	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
+	count := 0
+	var err error
+	logutil.BgLogger().Info(
+		"try to start runaway manager loop",
+		zap.Error(err))
+	for {
+		select {
+		case <-do.exit:
+			return
+		case <-runawayWatchSyncTicker.C:
+			// Due to the watch and watch done tables is created later than runaway queries table
+			err = do.updateNewAndDoneWatch()
+			if err == nil {
+				do.wg.Run(do.runawayRecordFlushLoop, "runawayRecordFlushLoop")
+				do.wg.Run(do.runawayWatchSyncLoop, "runawayWatchSyncLoop")
+				do.runawayManager.MarkSyncerInitialized()
+				return
+			}
+		}
+		count++
+		if count == 1800 {
+			logutil.BgLogger().Warn(
+				"failed to start runaway manager loop, please check whether the bootstrap or update is finished",
+				zap.Error(err))
+		}
+	}
+}
+
 func (do *Domain) updateNewAndDoneWatch() error {
 	do.runawaySyncer.mu.Lock()
 	defer do.runawaySyncer.mu.Unlock()
 	records, err := do.runawaySyncer.getNewWatchRecords()
 	if err != nil {
-		logutil.BgLogger().Error("try to get new runaway watch", zap.Error(err))
 		return err
 	}
 	for _, r := range records {
@@ -153,7 +184,6 @@ func (do *Domain) updateNewAndDoneWatch() error {
 	}
 	doneRecords, err := do.runawaySyncer.getNewWatchDoneRecords()
 	if err != nil {
-		logutil.BgLogger().Error("try to get done runaway watch", zap.Error(err))
 		return err
 	}
 	for _, r := range doneRecords {
@@ -165,6 +195,7 @@ func (do *Domain) updateNewAndDoneWatch() error {
 func (do *Domain) runawayWatchSyncLoop() {
 	defer util.Recover(metrics.LabelDomain, "runawayWatchSyncLoop", nil, false)
 	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
+	lastLogErrorTime := time.Now()
 	for {
 		select {
 		case <-do.exit:
@@ -172,7 +203,10 @@ func (do *Domain) runawayWatchSyncLoop() {
 		case <-runawayWatchSyncTicker.C:
 			err := do.updateNewAndDoneWatch()
 			if err != nil {
-				logutil.BgLogger().Warn("get runaway watch record failed", zap.Error(err))
+				if now := time.Now(); now.Sub(lastLogErrorTime) > runawayLoopLogErrorInterval {
+					logutil.BgLogger().Warn("get runaway watch record failed", zap.Error(err))
+					lastLogErrorTime = now
+				}
 			}
 		}
 	}

